quotesimulator/quote_simulator.go

```diff
@@ -0,0 +1,82 @@
+package quotesimulator
+
+import (
+	"context"
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/osmosis-labs/osmosis/osmomath"
+	"github.com/osmosis-labs/osmosis/v26/app/params"
+	txfeestypes "github.com/osmosis-labs/osmosis/v26/x/txfees/types"
+	"github.com/osmosis-labs/sqs/domain"
+	"github.com/osmosis-labs/sqs/domain/cosmos/auth/types"
+	"github.com/osmosis-labs/sqs/domain/cosmos/tx"
+
+	poolmanagertypes "github.com/osmosis-labs/osmosis/v26/x/poolmanager/types"
+)
+
+// quoteSimulator simulates a quote and returns the gas adjusted amount and the fee coin.
+type quoteSimulator struct {
+	msgSimulator       tx.MsgSimulator
+	encodingConfig     params.EncodingConfig
+	txFeesClient       txfeestypes.QueryClient
+	accountQueryClient types.QueryClient
+	chainID            string
+}
+
+func NewQuoteSimulator(msgSimulator tx.MsgSimulator, encodingConfig params.EncodingConfig, txFeesClient txfeestypes.QueryClient, accountQueryClient types.QueryClient, chainID string) *quoteSimulator {
+	return &quoteSimulator{
+		msgSimulator:       msgSimulator,
+		encodingConfig:     encodingConfig,
+		txFeesClient:       txFeesClient,
+		accountQueryClient: accountQueryClient,
+		chainID:            chainID,
+	}
+}
+
+// SimulateQuote implements domain.QuoteSimulator
+func (q *quoteSimulator) SimulateQuote(ctx context.Context, quote domain.Quote, slippageToleranceMultiplier osmomath.Dec, simulatorAddress string) (uint64, sdk.Coin, error) {
+	route := quote.GetRoute()
+	if len(route) != 1 {
+		return 0, sdk.Coin{}, fmt.Errorf("route length must be 1, got %d", len(route))
+	}
+
+	poolsInRoute := route[0].GetPools()
+
+	// Create the pool manager route
+	poolManagerRoute := make([]poolmanagertypes.SwapAmountInRoute, len(poolsInRoute))
+	for i, r := range poolsInRoute {
+		poolManagerRoute[i] = poolmanagertypes.SwapAmountInRoute{
+			PoolId:        r.GetId(),
+			TokenOutDenom: r.GetTokenOutDenom(),
+		}
+	}
+
+	// Slippage bound from the token in and provided slippage tolerance multiplier
+	tokenOutAmt := quote.GetAmountOut()
+	slippageBound := tokenOutAmt.ToLegacyDec().Mul(slippageToleranceMultiplier).TruncateInt()
+
+	// Create the swap message
+	swapMsg := &poolmanagertypes.MsgSwapExactAmountIn{
```

Comment:
Can it be that result can vary depending of swap method ( `poolmanagertypes.MsgSwapExactAmountIn` vs  `poolmanagertypes.MsgSwapExactAmountOut` ) for some cases such as imbalanced pools, etc?

Reply:
"in given out" is not supported in simulation for the same reason I mentioned here:
https://github.com/osmosis-labs/sqs/pull/547#discussion_r1828122218